src/Catel.Core/Catel.Core.Shared/Services/ObjectIdGenerator.cs

```diff
@@ -0,0 +1,63 @@
+﻿// --------------------------------------------------------------------------------------------------------------------
+// <copyright file="ObjectIdGenerator.cs" company="Catel development team">
+//   Copyright (c) 2008 - 2018 Catel development team. All rights reserved.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+
+namespace Catel.Services
+{
+    using System.Collections.Generic;
+
+    /// <summary>
+    /// The ObjectIdGenerator class.
+    /// </summary>
+    /// <typeparam name="TObjectType">The object type</typeparam>
+    /// <typeparam name="TUniqueIdentifier">The unique identifier type</typeparam>
+    public abstract class ObjectIdGenerator<TObjectType, TUniqueIdentifier> : IObjectIdGenerator<TObjectType, TUniqueIdentifier>
+        where TObjectType : class
+    {
+        private static readonly Queue<TUniqueIdentifier> _releasedUniqueIdentifiers = new Queue<TUniqueIdentifier>();
+
+        private static readonly object _syncObj = new object();
+
+        /// <inheritdoc />
+        public TUniqueIdentifier GetUniqueIdentifier()
+        {
+            lock (_syncObj)
+            {
+                return GenerateUniqueIdentifier();
+            }
+        }
+
+        /// <inheritdoc />
+        public TUniqueIdentifier GetUniqueIdentifier(bool reuse)
```

Comment:
Maybe it's better to add a default value for 'reuse' argument instead of overloading method?

Reply:
Eventually, the usage of default values would turn into a breaking change.

Reply:
So a default value of reuse = false would be fine?

Reply:
I think yes.

Reply:
It seems to me that this flag must be immutable. If true, we should store released identifiers and reuse them. If false, we just creating new ids. Otherwise queue will be grow on every instantiation, isn't it?

Reply:
Good point! We could even consider to remove the lock (we can simply lock the collection itself). Even further optimizing, we should only create the queue when the flag is true. This saves 2 objects / instance if not being used.

Reply:
Now I implemented this in this way:

            serviceLocator.RegisterType(typeof(IObjectIdGenerator<int>), x => typeof(IntegerObjectIdGenerator<>).MakeGenericTypeEx(type), type, RegistrationType.Singleton, false);
            _objectIdGenerator = serviceLocator.ResolveType<IObjectIdGenerator<int>>(type);
            UniqueIdentifier = _objectIdGenerator.GetUniqueIdentifier(reuseIdentifier);

To change this we need to do something like this:

            var typeFactory = serviceLocator.ResolveType<ITypeFactory>();
            serviceLocator.RegisterType(typeof(IObjectIdGenerator<int>), x => typeFactory.CreateInstanceWithParameters(typeof(IntegerObjectIdGenerator<>).MakeGenericTypeEx(type), reuseIdentifier), type, RegistrationType.Singleton, false);
            _objectIdGenerator = serviceLocator.ResolveType<IObjectIdGenerator<int>>(type);
            UniqueIdentifier = _objectIdGenerator.GetUniqueIdentifier(reuseIdentifier);

and also force the existence of constructor with the bool argument for custom generators.

Reply:
What if a custom generator have a bool argument that means anything but reuse?

Reply:
Fixed queue allocation

Reply:
IMHO, as long as this flag is mutable, we need to think about what behavior should be when changing the value.
Can we make it immutable by setting it through a public property? To do this, we can add "if" statement in the property setter to check whether it was initialized and (or) Id generation was started already. It's a bit ugly, but i think better then ctor injection through TypeFactory.

Reply:
OK, you convinced me :)